nft-simple/src/nft_core.rs

```diff
@@ -133,40 +144,78 @@ impl NonFungibleTokenCore for Contract {
     }
 
     #[payable]
-    fn nft_approve_account_id(&mut self, token_id: TokenId, account_id: ValidAccountId) -> bool {
-        assert_one_yocto();
-        let mut token = self.tokens_by_id.get(&token_id).expect("Token not found");
-        assert_eq!(&env::predecessor_account_id(), &token.owner_id);
+    fn nft_approve(
+        &mut self,
+        token_id: TokenId,
+        account_id: ValidAccountId,
+        msg: Option<String>,
+    )  -> Option<Promise> {
         let account_id: AccountId = account_id.into();
-        let storage_used = bytes_for_approved_account_id(&account_id);
-        if token.approved_account_ids.insert(account_id) {
-            deposit_refund(storage_used);
+        let storage_used = bytes_for_approved_account_id((&account_id, &0_u64));
+
+        let mut token = self.tokens_by_id.get(&token_id).expect("Token not found");
+
+        if token.approved_account_ids.contains_key(&account_id) {
+            // If account is already approved, refund the attached deposit
+            if env::attached_deposit() != 0 {
+                Promise::new(env::predecessor_account_id()).transfer(env::attached_deposit());
+            }
+        } else {
+            assert!(env::attached_deposit() >= storage_used as u128, "attached_deposit doesn't cover storage of account_id: {}", account_id.clone());
+        }
+
+        assert_eq!(&env::predecessor_account_id(), &token.owner_id, "Predecessor must be the token owner.");
+
+        token.approval_counter += 1;
+
+        let mut account_approval_id = HashMap::new();
+        account_approval_id.insert(account_id.clone(), token.approval_counter);
+
+        if token.approved_account_ids.insert(account_id.clone(), token.approval_counter).is_some() {
             self.tokens_by_id.insert(&token_id, &token);
-            true
+            deposit_refund(storage_used);
+        }
+
+        self.tokens_by_id.insert(&token_id, &token);
+
+        if let Some(msg) = msg {
+            Some(ext_non_fungible_approval_receiver::nft_on_approve(
+                token_id,
+                token.owner_id,
+                U64::from(token.approval_counter),
+                msg,
+                &account_id,
+                NO_DEPOSIT,
+                env::prepaid_gas() - GAS_FOR_NFT_TRANSFER_CALL,
+            ))
```

Comment:
Seems you can use `.as_return()` method to mark it as returned. https://docs.rs/near-sdk/3.0.1/near_sdk/struct.Promise.html#method.as_return

```suggestion
           ext_non_fungible_approval_receiver::nft_on_approve(
                token_id,
                token.owner_id,
                U64::from(token.approval_counter),
                msg,
                &account_id,
                NO_DEPOSIT,
                env::prepaid_gas() - GAS_FOR_NFT_TRANSFER_CALL,
            ).as_return()
```

Reply:
Hm, doesn't build for me:
```rust
182 | /             ext_non_fungible_approval_receiver::nft_on_approve(
183 | |                 token_id,
184 | |                 token.owner_id,
185 | |                 U64::from(token.approval_counter),
...   |
189 | |                 env::prepaid_gas() - GAS_FOR_NFT_TRANSFER_CALL,
190 | |             ).as_return()
    | |_________________________^ expected enum `std::option::Option`, found struct `near_sdk::Promise`
```

Reply:
ah wait, saw your comment above this one

Reply:
do we need this? We're not going to use return val for nft_approve.

Reply:
This way you can know if your token got listed by just listening to the promise